spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,503 @@
+    result := VirtualNetworksSubnetsSpecArm{}
+    
+    // Copy properties
+    result.Properties = o.Properties
+    result.Type = o.Type
+    
+    // Name is special
+    result.Name = CreateFullyQualifiedArmName(parentArmId, o.Name)
+    return result, nil
+}
+
+// Example usage
+func ExamplePut() error {
+    etcdObject := MagicallyGetStateFromEtcd()
+    
+    // Somehow we determine that this particular object has an owner - possibly we need to impl an interface?
+    ownerArmId = etcdObject.FQArmId()
+
+    var toSerialize interface{}
+    toSerialize = etcdObject
+    if armTransformer, ok := etcdObject.(ARMTransformer); ok {
+        toSerialize = armTransformer.ToArm(ownerArmId)
+    }
+    json := json.Marshal(toSerialize)
+    armClient.SendIt(json)
+
+}
+
+func ExampleGet() error {
+    etcdObject := MagicallyGetStateFromEtcd()
+    
+    // TODO: Complete this
+
+    // Somehow we determine that this particular object has an owner - possibly we need to impl an interface?
+    ownerArmId = etcdObject.Owner.FQArmId()
+    
+    armId := etcdObject.FQArmId()
+
+    // We need to provide the empty type to deserialize into
+    // Somehow construct a new object of type etcdObject
+    if armTransformer, ok := newEtcdObject.(ARMTransformer); ok {
+        toGet = armTransformer.ToArm(ownerArmId) // This just converts from an empty etcdObject shape to an empty arm object shape
+        armClient.GetIt(armId, toGet)
+    
+        armTransformer.FromArm(toGet)
+    }
+}
+```
+
+### A note on the differences between owner references and arbitrary resource references
+Because we are code-generating all of the `ownerRef` fields, we can always supply the annotations for group and kind automatically, using the information
+provided to us by the ARM json specification. **This is not the case for abitrary references (`id`'s) to other resources**. We do not actually know
+programmatically what type that reference is, and it some cases it may actually be allowed to point to multiple different types 
+(think custom image vs shared image gallery).
+
+Currently, the plan is to trust the customer gets it right and ensure we have good
+failure modes for resources if they actually got it wrong (which will cause their resource to move to an unhealthy state). 
+
+## FAQ
+
+#### What happens when a dependent resource specifies an `ownerRef` that doesn't exist?
+The dependent resource will be stuck in an unprovisioned state with an error stating that the owner doesn't exist.
+If the owner is created, the dependent resource will also be created automatically by the reconciliation loop.
+
+#### What happens when a resource contains a link to another resource which doesn't exist?
+The resource with the link will be stuck in an unprovisioned state with an error stating that the linked resource doesn't exist.
+This behavior is the same as for a dependent resource with a non-existant owner.
```

Comment:
Might this result in deadlock if two resources reference each other? We might need to know which one to create first.

Reply:
If that was allowed by ARM, yes that could be an issue. I don't think that's allowed by ARM... will do some reading to see if I can find official documentation on that.